keras_maskrcnn/preprocessing/csv_generator.py

```diff
@@ -0,0 +1,189 @@
+    if sys.version_info[0] < 3:
+        return open(path, 'rb')
+    else:
+        return open(path, 'r', newline='')
+
+
+class CSVGenerator(Generator):
+    def __init__(
+        self,
+        csv_data_file,
+        csv_class_file,
+        base_dir=None,
+        **kwargs
+    ):
+        self.image_names = []
+        self.image_data  = {}
+        self.base_dir    = base_dir
+
+        # Take base_dir from annotations file if not explicitly specified.
+        if self.base_dir is None:
+            self.base_dir = os.path.dirname(csv_data_file)
+
+        # parse the provided class file
+        try:
+            with _open_for_csv(csv_class_file) as file:
+                self.classes = _read_classes(csv.reader(file, delimiter=','))
+        except ValueError as e:
+            raise_from(ValueError('invalid CSV class file: {}: {}'.format(csv_class_file, e)), None)
+
+        self.labels = {}
+        for key, value in self.classes.items():
+            self.labels[value] = key
+
+        # csv with img_path, x1, y1, x2, y2, class_name
+        try:
+            with _open_for_csv(csv_data_file) as file:
+                self.image_data = _read_annotations(csv.reader(file, delimiter=','), self.classes)
+        except ValueError as e:
+            raise_from(ValueError('invalid CSV annotations file: {}: {}'.format(csv_data_file, e)), None)
+        self.image_names = list(self.image_data.keys())
+
+        super(CSVGenerator, self).__init__(**kwargs)
+
+    def size(self):
+        return len(self.image_names)
+
+    def num_classes(self):
+        return max(self.classes.values()) + 1
+
+    def name_to_label(self, name):
+        return self.classes[name]
+
+    def label_to_name(self, label):
+        return self.labels[label]
+
+    def image_path(self, image_index):
+        return os.path.join(self.base_dir, self.image_names[image_index])
+
+    def image_aspect_ratio(self, image_index):
+        # PIL is fast for metadata
+        image = Image.open(self.image_path(image_index))
+        return float(image.width) / float(image.height)
+
+    def load_image(self, image_index):
+        return read_image_bgr(self.image_path(image_index))
+
+    def load_annotations(self, image_index):
+        path   = self.image_names[image_index]
+        annots = self.image_data[path]
+
+        # find mask size in order to allocate the right dimension for the annotations
+        annotations  = np.zeros((len(annots), 5))
+        masks        = []
+
+        for idx, annot in enumerate(annots):
+            annotations[idx, 0]  = float(annot['x1'])
+            annotations[idx, 1]  = float(annot['y1'])
+            annotations[idx, 2]  = float(annot['x2'])
+            annotations[idx, 3]  = float(annot['y2'])
+            annotations[idx, 4]  = self.name_to_label(annot['class'])
+            mask = cv2.imread(annot['mask_path'], cv2.IMREAD_GRAYSCALE).astype(float)
+            mask[np.where(mask > 0)] = 1  # convert from 0-255 to binary mask
```

Comment:
This can be simplified to `mask = (mask > 0).astype(np.uint8)`.

Reply:
done.